File: kimdaeyeong/src/baekjoon/Boj9177BFS.java
package baekjoon;

import java.util.*;
import java.io.*;

/**
 * 백준 9177
 * 단어 섞기
 * 골드4
 * https://www.acmicpc.net/problem/9177
 */
public class Boj9177BFS {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        for(int t = 1; t <= T; t++) {
            st = new StringTokenizer(br.readLine());

            char[] s1 = st.nextToken().toCharArray();
            char[] s2 = st.nextToken().toCharArray();
            char[] s3 = st.nextToken().toCharArray();

            sb.append("Data set ").append(t).append(": ").append(bfs(s1, s2, s3)).append("\n");
        }

        System.out.print(sb);
    }

    public static String bfs(char[] s1, char[] s2, char[] s3) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[201][201];

        q.add(new int[] {0, 0, 0});
        visited[0][0] = true;

        while(!q.isEmpty()) {
            int[] cur = q.poll();

            if(cur[2] == s3.length) {
                return "yes";
            }

            if(cur[0] < s1.length && !visited[cur[0] + 1][cur[1]] &&
                    s1[cur[0]] == s3[cur[2]]) {
                q.add(new int[] {cur[0] + 1, cur[1], cur[2] + 1});
                visited[cur[0] + 1][cur[1]] = true;
            }

            if(cur[1] < s2.length && !visited[cur[0]][cur[1] + 1] &&
                    s2[cur[1]] == s3[cur[2]]) {
                q.add(new int[] {cur[0], cur[1] + 1, cur[2] + 1});
                visited[cur[0]][cur[1] + 1] = true;
            }
        }

        return "no";
    }

}

File: limsubin/baekjoon/boj9177_bfs.py
# BOJ 9177: 단어 섞기
# https://www.acmicpc.net/problem/9177

import sys
from collections import deque

def input():
    return sys.stdin.readline().strip()

def bfs(a, b, c):
    queue = deque()
    queue.append([0, 0])

    size_a = len(a)
    size_b = len(b)
    size_c = len(c)

    # 방문한 a, b 의 길이
    visited = [[False] * (size_b + 1) for _ in range(size_a + 1)]
    visited[0][0] = True

    while queue:
        x, y = queue.popleft()

        # c의 다음 문자가 a의 다음 문자와 일치할 때
        if x+1 <= size_a and not visited[x+1][y] and c[x + y] == a[x]:
            # c 탐색 완료
            if x + y + 1 == size_c:
                return "yes"
            # 방문 체크 및 큐에 삽입
            visited[x+1][y] = True
            queue.append([x+1, y])

        # c의 다음 문자가 b의 다음 문자와 일치할 때
        if y+1 <= size_b and not visited[x][y+1] and c[x + y] == b[y]:
            # c 탐색 완료
            if x + y + 1 == size_c:
                return "yes"
            # 방문 체크 및 큐에 삽입
            visited[x][y+1] = True
            queue.append([x, y+1])

    return "no"

N = int(input())

for n in range(1, N+1):
    a, b, c = input().split()
    print(f'Data set {n}: {bfs(a, b, c)}')
